backend/src/main/java/org/eclipse/tractusx/puris/backend/common/edc/logic/service/EdcAdapterService.java

```diff
@@ -487,6 +543,379 @@ public Response getProxyPullRequest(String url, String authKey, String authCode,
+        };
+        boolean failed = true;
+        EdcContractMapping contractMapping = edcContractMappingService.find(partner.getBpnl());
+        try {
+            if (contractMapping != null) {
+                String itemStockContractId = contractMapping.getItemStockContractId();
+                String assetId = contractMapping.getItemStockAssetId();
+                String partnerItemStockEdcUrl = contractMapping.getItemStockEdcProtocolUrl();
+                if (itemStockContractId == null || assetId == null || partnerItemStockEdcUrl == null) {
+                    log.info("Need Contract for ItemStock 2.0.0");
+                    if (negotiateForPartnerItemStock2(mpr, direction)) {
+                        contractMapping = edcContractMappingService.find(partner.getBpnl());
+                        itemStockContractId = contractMapping.getItemStockContractId();
+                        assetId = contractMapping.getItemStockEdcProtocolUrl();
+                        partnerItemStockEdcUrl = contractMapping.getItemStockEdcProtocolUrl();
+                    } else {
+                        log.error("Failed to contract for ItemStock 2.0.0 with " + partner.getBpnl());
+                        return null;
+                    }
+                }
+                try {
+                    if (!partnerItemStockEdcUrl.equals(partner.getEdcUrl())){
+                        log.warn("Diverging EDC URL's for partner " + partner.getBpnl());
+                        log.warn("Edc URL from AAS " + partnerItemStockEdcUrl);
+                        log.warn("General Edc Url of partner " + partner.getEdcUrl());
+                    }
+                    // Request EdrToken
+                    var transferResp = initiateProxyPullTransfer(partner, itemStockContractId, assetId, partnerItemStockEdcUrl);
+                    String transferId = transferResp.get("@id").asText();
+                    for (int i = 0; i < 100; i++) {
+                        Thread.sleep(100);
+                        transferResp = getTransferState(transferId);
+                        if ("STARTED".equals(transferResp.get("edc:state").asText())) {
+                            break;
+                        }
+                    }
+                    EDR_Dto edr_Dto = null;
+                    // Await arrival of edr
+                    for (int i = 0; i < 100; i++) {
+                        Thread.sleep(100);
+                        edr_Dto = edrService.findByTransferId(transferId);
+                        if (edr_Dto != null) {
+                            log.info("Successfully negotiated for " + assetId + " with " + partner.getEdcUrl());
+                            break;
+                        }
+                    }
+
+                    if (edr_Dto != null) {
+                        try (var response = getProxyPullRequest(edr_Dto.endpoint(), edr_Dto.authKey(), edr_Dto.authCode(), new String[]{materialNumber, direction.toString()})) {
```

Comment:
The submodel href-field now contains the cx id and direction. This href is now actually being used in the EdcAdapterService for an ItemStock Submodel request. Additionally, a warning is written to the logs, if the URL from the href diverges from the URL, we would otherwise generate from the EDR token.